### ISSUE TITLE
makedirs fails some times in zipstream

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3625]
|<img alt="msyang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| msyang reported|
|-|-|
|Trac ID|trac#3625|
|Type|defect|
|Created|2009-01-18 21:54:28Z|
|Branch|https://github.com/twisted/twisted/tree/zipstream-directory-3625|

In:

[twisted/python/zipstream.py#L362@22464](twisted/python/zipstream.py#L362@22464)

```python
os.makedirs(f)
```

should be changed to:

```python
os.makedirs(fdir)
```

If the zipfile doesn't first explicitly specify a directory before specifying a file in that directory, the os.makedirs(f) will make a directory having the filename 'f' and then not overwrite it with the actual file contents.

to test:

/mysampledirectory
   myfile.txt

 * make a zipfile consisting only of /mysampledirectory/myfile.txt.  Let's call it myzip.zip

```python
import os
from twisted.python.zipstream import unzip
unzip('myzip.zip')
assert not os.path.isdir('mysampledirectory/myfile.txt')
# will fail
```

<details><summary>Searchable metadata</summary>

```
trac-id__3625 3625
type__defect defect
reporter__msyang msyang
priority__normal normal
milestone__ 
branch__branches_zipstream_directory_3625 branches/zipstream-directory-3625
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__zipstream__python zipstream, python
time__1232315668000000 1232315668000000
changetime__1232936160000000 1232936160000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [26086]) Branching to 'zipstream-directory-3625'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_3"></a>|@mwhudson set owner to @exarkun|
|-|-|

Looks fine.  Merge if you like, but...

As usual <wink> I have some comments about the tests.  Naming the files after the integer sequence number in makeZipFile is pretty obscure, and the repeated verification of the extracted tree contents is, well, repetitive.  Something like bzr's build_tree_contents interface:

http://starship.python.net/crew/mwh/bzrlibapi/bzrlib.tests.TestCaseInTempDir.html#build_tree_contents

would make the tests much clearer, I think (together with a verify_tree_contents helper).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

I agree that `makeZipFile` is pretty ghetto, and an API like the one you point out in bzrlib would be helpful in several places in Twisted's test suite.  I'll figure out one or more tickets to file for dealing with those points.  For now, I'll merge this, since these issues are mainly pre-existing in the zipstream test code.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [26122]) Python 2.3 compatibility

refs [#3625](https://github.com/twisted/twisted/issues/3625)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set status to `closed`|
|-|-|

(In [26124]) Merge zipstream-directory-3625

Author: exarkun
Reviewer: mwhudson
Fixes: [#3625](https://github.com/twisted/twisted/issues/3625)

Fix the previous misbehavior of `twisted.python.zipstream` which led it
to extract the first file in a directory in a zipfile as an empty directory
rather than as a file with the proper contents.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
